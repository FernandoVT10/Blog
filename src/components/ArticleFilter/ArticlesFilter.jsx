import "./ArticlesFilter.scss";

export default () => {
    return (
        <div className="articles-filter">
            <form>
                <div className="articles-filter__search-container">
                    <input
                    className="articles-filter__search-input"
                    type="search"
                    placeholder="Search article"
                    autoComplete="search" />

                    <button type="submit" className="articles-filter__search-button">
                        <i className="fas fa-search" aria-hidden="true"></i>
                    </button>
                </div>
                
                <a href="#" className="articles-filter__filter-button">
                    <i className="fas fa-sort-amount-down-alt" aria-hidden="true"></i>
                </a>
            </form>
        </div>
    );
};