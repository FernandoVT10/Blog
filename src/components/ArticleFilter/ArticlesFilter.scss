@use "../../styles/variables";

.articles-filter {
    &__search-container {
        position: relative;
        display: inline-block;
        width: calc(100% - 30px);
        position: relative;
    }

    &__search-button {
        outline: 0;
        position: absolute;
        top: 12px;
        right: 10px;
        background: transparent;
        border: none;
        color: white;
    }

    &__filter-button {
        outline: 0;
        margin-left: 10px;
        display: inline-block;
        background: transparent;
        border: none;
        color: white;
        transition: .3s;

        &:hover,
        &--active {
            color: variables.$secondary;
        }
    }
    
    &__filter {
        position: absolute;
        display: none;
        z-index: 1000;
        right: 0;
        top: 50px;
        width: 90vw;
        max-width: 600px;
        background: variables.$black;

        &--active {
            display: block;
        }
    }

    &__filter-search {
        display: none;
    }

    &__categories-container {
        padding: 15px;
    }

    &__categories-title {
        color: white;
        font-weight: bold;
        font-size: 18px;
    }

    &__category {
        margin-right: 10px;
        display: inline;
    }

    &__checkbox {
        display: none;
    }

    &__label {
        cursor: pointer;
        font-weight: 600;
        font-size: 14px;
        color: white;
        transition: .3s;

        &:hover {
            color: #cccccc;
        }
    }

    &__label-checkbox {
        position: relative;
        top: 3px;
        display: inline-block;
        width: 15px;
        height: 15px;
        border: solid 3px variables.$secondary;
        margin-right: 5px;
    }

    &__checkbox:checked ~ &__label {
        & .articles-filter__label-checkbox {
            background: variables.$secondary;

            &:after {
                content: "";
                position: absolute;
                top: -2px;
                left: 3px;
                width: 5px;
                height: 10px;
                border: solid white;
                border-width: 0 2px 2px 0;
                -webkit-transform: rotate(45deg);
                -ms-transform: rotate(45deg);
                transform: rotate(45deg);
            }
        }
    }

    &__set-categories-button {
        display: block;
        margin-top: 10px;
        float: right;
    }

    @media screen and (max-width: 576px) {
        &__search-container {
            display: none;
        }

        &__filter-search {
            display: block;
        }

        &__filter-button {
            float: right;
        }
    }
}