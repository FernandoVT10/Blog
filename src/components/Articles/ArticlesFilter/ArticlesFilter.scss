.articles-filter {
    position: relative;

    &__search-container {
        position: relative;
        display: inline-block;
        width: calc(100% - 30px);
        position: relative;
    }

    &__search-button {
        outline: 0;
        position: absolute;
        top: 12px;
        right: 10px;
        background: transparent;
        border: none;
        color: var(--white);
    }

    &__filter-button {
        outline: 0;
        margin-left: 10px;
        display: inline-block;
        background: transparent;
        border: none;
        color: var(--white);
        transition: .3s;

        &:hover,
        &--active {
            color: var(--secondary);
        }
    }
    
    &__filter {
        position: absolute;
        display: none;
        z-index: 1000;
        right: 0;
        top: 50px;
        padding-bottom: 10px;
        width: 90vw;
        max-width: 600px;
        background: var(--black);

        &--active {
            display: block;
        }
    }

    &__filter-search {
        display: none;
    }

    &__categories-container {
        padding: 15px;
    }

    &__categories-title {
        color: var(--white);
        font-weight: bold;
        font-size: 18px;
    }

    &__set-categories-button {
        float: right;
        display: block;
        padding: 10px 20px;
        font-weight: 600;
        border: none;
        color: var(--black);
        background: var(--secondary);
        margin-top: 10px;
        transition: .3s;

        &:hover {
            color: var(--secondary);
            background: var(--black);
        }
    }

    @media screen and (max-width: 576px) {
        &__filter {
            top: 20px;
            right: -10px;
        }

        &__search-container {
            display: none;
        }

        &__filter-search {
            display: block;
        }

        &__filter-button {
            float: right;
        }
    }
}