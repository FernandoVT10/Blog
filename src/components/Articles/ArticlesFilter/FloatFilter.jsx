export default ({ search, setSearch, setCategories, categories, filterActive }) => {
    const handleCategory = name => {
        setCategories(categories.map(category => {
            if(category.name === name) {
                Object.assign(category, { active: !category.active });
            }

            return category;
        }));
    };

    const getCategories = () => {
        return categories.map(category => {
            return (
                <div key={category._id} className="formulary__checkbox">
                    <input
                    className="formulary__checkbox__input"
                    id={`category-${category.name}`}
                    onChange={() => handleCategory(category.name)}
                    type="checkbox"
                    checked={category.active} />

                    <label
                    className="formulary__checkbox__label"
                    htmlFor={`category-${category.name}`}>
                        <span className="formulary__checkbox__label-checkbox"></span>
                        { category.name }
                    </label>
                </div>
            );
        });
    };

    const filterClass = filterActive ? "articles-filter__filter--active" : "";

    return (
        <div className={`articles-filter__filter ${filterClass}`}>
            <input
            className="formulary__input articles-filter__filter-search"
            type="search"
            placeholder="Search article"
            value={search}
            onChange={({ target: { value } }) => setSearch(value)}
            autoComplete="search" />

            <div className="articles-filter__categories-container">
                <h4 className="articles-filter__categories-title">Categories</h4>

                <div className="aticles-filter__categories">
                    { getCategories() }
                </div>

                <button
                type="submit"
                className="articles-filter__set-categories-button">
                    Set Categories
                </button>
            </div>
        </div>
    );
}