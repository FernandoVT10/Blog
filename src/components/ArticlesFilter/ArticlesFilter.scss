@use "../../styles/variables";

.articles-filter {
    position: relative;

    &__search-container {
        position: relative;
        display: inline-block;
        width: calc(100% - 30px);
        position: relative;
    }

    &__search-button {
        outline: 0;
        position: absolute;
        top: 12px;
        right: 10px;
        background: transparent;
        border: none;
        color: white;
    }

    &__filter-button {
        outline: 0;
        margin-left: 10px;
        display: inline-block;
        background: transparent;
        border: none;
        color: white;
        transition: .3s;

        &:hover,
        &--active {
            color: variables.$secondary;
        }
    }
    
    &__filter {
        position: absolute;
        display: none;
        z-index: 1000;
        right: 0;
        top: 50px;
        padding-bottom: 10px;
        width: 90vw;
        max-width: 600px;
        background: variables.$black;

        &--active {
            display: block;
        }
    }

    &__filter-search {
        display: none;
    }

    &__categories-container {
        padding: 15px;
    }

    &__categories-title {
        color: white;
        font-weight: bold;
        font-size: 18px;
    }

    &__set-categories-button {
        display: block;
        margin-top: 10px;
        float: right;
    }

    @media screen and (max-width: 576px) {
        &__filter {
            top: 20px;
            right: -10px;
        }

        &__search-container {
            display: none;
        }

        &__filter-search {
            display: block;
        }

        &__filter-button {
            float: right;
        }
    }
}