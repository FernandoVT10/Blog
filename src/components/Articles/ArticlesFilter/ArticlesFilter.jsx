import { useState, useEffect, useRef } from "react";
import { useRouter } from "next/router";
import Api from "../../../ApiController";

import "./ArticlesFilter.scss";

function ArticlesFilter() {
    const [filterActive, setFilterActive] = useState(false);
    const [search, setSearch] = useState("");
    const [categories, setCategories] = useState([]);

    const router = useRouter();

    // we active the categories with the URL query
    const setCategoriesStatus = (categories, queryCategories) => {
        setCategories(categories.map(({ _id, name }) => {
            if(queryCategories.includes(name)) {
                return { active: true, _id, name };
            } else {
                return { active: false, _id, name };
            }
        }));
    }

    useEffect(() => {
        // we use the URLSearchParams with window.location.search, 
        // because the router.query is undefined in the first call
        const queryCategories = new URLSearchParams(location.search).getAll("categories");

        Api.get("categories/getAllCategories/")
        .then(categories => setCategoriesStatus(categories, queryCategories));
    }, []);

    useEffect(() => {
        if(router.query.search) {
            setSearch(router.query.search);
        }
    }, [router]);

    const handleForm = e => {
        e.preventDefault();
        setFilterActive(false);

        const activeCategories = [];
        const query = router.query;

        delete query.search;
        delete query.categories;

        categories.forEach(({ active, name }) => {
            if(active) {
                activeCategories.push(name);
            }
        });

        if(search) {
            query.search = search;
        }

        if(activeCategories.length) {
            query.categories = activeCategories;
        }

        router.push({
            pathname: router.pathname,
            query: query
        });
    };

    // Active or desactive the category
    const handleCategory = name => {
        setCategories(categories.map(category => {
            if(category.name === name) {
                Object.assign(category, { active: !category.active });
            }

            return category;
        }));
    };

    const getCategories = () => {
        if(categories.length) {
            return categories.map(category => {
                return (
                    <div key={category._id} className="formulary__checkbox">
                        <input
                        className="formulary__checkbox__input"
                        id={`category-${category.name}`}
                        onChange={() => handleCategory(category.name)}
                        type="checkbox"
                        checked={category.active} />

                        <label
                        className="formulary__checkbox__label"
                        htmlFor={`category-${category.name}`}>
                            <span className="formulary__checkbox__label-checkbox"></span>
                            { category.name }
                        </label>
                    </div>
                );
            });
        }
    };

    const filterClass = filterActive ? "articles-filter__filter--active" : "";
    const filterButtonClass = filterActive ? "articles-filter__filter-button--active" : "";
    
    return (
        <div className="articles-filter">
            <form onSubmit={handleForm}>
                <div className="articles-filter__search-container">
                    <input
                    className="formulary__input"
                    type="search"
                    placeholder="Search article"
                    value={search}
                    onChange={({ target: { value } }) => setSearch(value)}
                    autoComplete="search" />

                    <button type="submit" className="articles-filter__search-button">
                        <i className="fas fa-search" aria-hidden="true"></i>
                    </button>
                </div>

                <div className={`articles-filter__filter ${filterClass}`}>
                    <input
                    className="formulary__input articles-filter__filter-search"
                    type="search"
                    placeholder="Search article"
                    value={search}
                    onChange={({ target: { value } }) => setSearch(value)}
                    autoComplete="search" />

                    <div className="articles-filter__categories-container">
                        <h4 className="articles-filter__categories-title">Categories</h4>

                        <div className="aticles-filter__categories">
                            { getCategories() }
                        </div>

                        <button
                        type="submit"
                        className="submit-button articles-filter__set-categories-button">
                            Set Categories
                        </button>
                    </div>
                </div>
                
                <a
                onClick={() => setFilterActive(!filterActive)}
                href="#"
                className={`articles-filter__filter-button ${filterButtonClass}`}>
                    <i className="fas fa-sort-amount-down-alt" aria-hidden="true"></i>
                </a>
            </form>
        </div>
    );
};

export default ArticlesFilter;