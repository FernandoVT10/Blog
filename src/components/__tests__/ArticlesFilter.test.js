import { act, Simulate } from "react-dom/test-utils";
import { render } from "react-dom";
import ArticlesFilter from "../ArticlesFilter/";
import { useRouter } from "next/router";

const CATEGORIES_MOCK = [
    {
        _id: 2,
        name: "Technology"
    },
    {
        _id: 6,
        name: "Landscapes"
    },
    {
        _id: 19,
        name: "Travels"
    }
];

let container;

describe('<ArticleFilter/> Component', () => {
    beforeEach(() => {
        fetchMock.doMock();

        fetchMock.mockOnce(JSON.stringify(
            CATEGORIES_MOCK
        ));

        container = document.createElement('div');
        document.body.appendChild(container);
    });
    
    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    it('It should active and desactive category', async () => {
        await act(async () => {
            render(<ArticlesFilter/>, container);
        });

        const checkbox = container.querySelector("#category-Technology");

        act(() => {
            Simulate.change(checkbox);
        });

        expect(checkbox.checked).toBeTruthy();

        act(() => {
            Simulate.change(checkbox);
        });
        expect(checkbox.checked).toBeFalsy();
    });

    it('It should change the search inputs value with query search value', async () => {
        useRouter.mockImplementation(() => ({
            query: { search: "Test value" }
        }));

        await act(async () => {
            render(<ArticlesFilter/>, container);
        });

        const inputsSearch = container.querySelectorAll(".formulary__input");

        inputsSearch.forEach(({ value }) => {
            expect(value).toBe("Test value")
        });
    });

    it('It should render the categories with an active', async () => {
        Object.defineProperty(window, 'location', {
            value: {
                search: "?categories=Landscapes"
            }
        });

        await act(async () => {
            render(<ArticlesFilter/>, container);
        });
        
        const category = container.querySelector("#category-Landscapes");

        expect(category.checked).toBeTruthy();
    });

    it('It should call router.push', async () => {
        Object.defineProperty(window, 'location', {
            value: {
                search: "?categories=Technology"
            }
        });

        const routerPush = jest.fn();

        useRouter.mockImplementation(() => ({
            query: { search: "Test value" },
            pathname: "/",
            push: routerPush
        }));

        await act(async () => {
            render(<ArticlesFilter/>, container);
        });

        const form = container.querySelector("form");

        act(() => {
            Simulate.submit(form);
        });

        expect(routerPush).toHaveBeenCalledWith({
            pathname: "/",
            query: {
                categories: ["Landscapes"],
                search: "Test value"
            }
        });
    });

    it('It should active and desactive the filter', async () => {
        await act(async () => {
            render(<ArticlesFilter/>, container);
        });

        const button = container.querySelector(".articles-filter__filter-button");
        const filter = container.querySelector(".articles-filter__filter");

        act(() => {
            Simulate.click(button);
        });

        expect(filter.classList.contains("articles-filter__filter--active")).toBeTruthy();
        expect(button.classList.contains("articles-filter__filter-button--active")).toBeTruthy();

        act(() => {
            Simulate.click(button);
        });

        expect(filter.classList.contains("articles-filter__filter--active")).toBeFalsy();
        expect(button.classList.contains("articles-filter__filter-button--active")).toBeFalsy();
    });
});