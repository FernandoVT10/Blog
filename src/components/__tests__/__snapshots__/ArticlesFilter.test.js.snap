// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleFilter/> Component Check if renders correctly 1`] = `
<div>
  <div
    class="articles-filter"
  >
    <form>
      <div
        class="articles-filter__search-container"
      >
        <input
          autocomplete="search"
          class="formulary__input"
          placeholder="Search article"
          type="search"
          value=""
        />
        <button
          class="articles-filter__search-button"
          type="submit"
        >
          <i
            aria-hidden="true"
            class="fas fa-search"
          />
        </button>
      </div>
      <div
        class="articles-filter__filter "
      >
        <input
          autocomplete="search"
          class="formulary__input articles-filter__filter-search"
          placeholder="Search article"
          type="search"
          value=""
        />
        <div
          class="articles-filter__categories-container"
        >
          <h4
            class="articles-filter__categories-title"
          >
            Categories
          </h4>
          <div
            class="aticles-filter__categories"
          >
            <div
              class="articles-filter__category"
            >
              <input
                class="articles-filter__checkbox"
                id="category-Technology"
                type="checkbox"
              />
              <label
                class="articles-filter__label"
                for="category-Technology"
              >
                <span
                  class="articles-filter__label-checkbox"
                />
                Technology
              </label>
            </div>
            <div
              class="articles-filter__category"
            >
              <input
                class="articles-filter__checkbox"
                id="category-Landscapes"
                type="checkbox"
              />
              <label
                class="articles-filter__label"
                for="category-Landscapes"
              >
                <span
                  class="articles-filter__label-checkbox"
                />
                Landscapes
              </label>
            </div>
            <div
              class="articles-filter__category"
            >
              <input
                class="articles-filter__checkbox"
                id="category-Travels"
                type="checkbox"
              />
              <label
                class="articles-filter__label"
                for="category-Travels"
              >
                <span
                  class="articles-filter__label-checkbox"
                />
                Travels
              </label>
            </div>
          </div>
          <button
            class="submit-button articles-filter__set-categories-button"
            type="submit"
          >
            Set Categories
          </button>
        </div>
      </div>
      <a
        class="articles-filter__filter-button "
        href="#"
      >
        <i
          aria-hidden="true"
          class="fas fa-sort-amount-down-alt"
        />
      </a>
    </form>
  </div>
</div>
`;
